packages/ERTP/src/issuer.js

```diff
@@ -102,39 +100,29 @@ function makeIssuerKit(allegedName, amountMathKind = MathKind.NAT) {
           const srcPaymentBalance = paymentLedger.get(srcPayment);
           // Note: this does not guarantee that optAmount itself is a valid stable amount
           assertAmountEqual(srcPaymentBalance, optAmount);
-          const purseBalance = purse.getCurrentAmount();
           const newPurseBalance = amountMath.add(
             srcPaymentBalance,
-            purseBalance,
+            currentBalance,
           );
           // Commit point
-          // eslint-disable-next-line no-use-before-define
-          const payments = reallocate(
-            harden({
-              payments: [srcPayment],
-              purses: [purse],
-              newPurseBalances: [newPurseBalance],
-            }),
-          );
-          assert(payments.length === 0, 'no payments should be returned');
+          // Move the assets in `srcPayment` into this purse, using up the
+          // source payment, such that total assets are conserved.
+          paymentLedger.delete(srcPayment);
+          currentBalance = newPurseBalance;
           return srcPaymentBalance;
         },
         withdraw: amount => {
           amount = amountMath.coerce(amount);
-          const purseBalance = purse.getCurrentAmount();
-          const newPurseBalance = amountMath.subtract(purseBalance, amount);
+          const newPurseBalance = amountMath.subtract(currentBalance, amount);
+          const payment = makePayment();
           // Commit point
-          // eslint-disable-next-line no-use-before-define
-          const [payment] = reallocate(
-            harden({
-              purses: [purse],
-              newPurseBalances: [newPurseBalance],
-              newPaymentBalances: [amount],
-            }),
-          );
+          // Move the withdrawn assets from this purse into a new payment
+          // which is returned. Total assets must remain conserved.
+          currentBalance = newPurseBalance;
+          paymentLedger.init(payment, amount);
```

Comment:
What's the reasoning for reversing the order of the update to `currentBalance`? Is it always erring on the least amount of rights?

Reply:
Yes, exactly! If we somehow fail in the middle, rights are magically dropped rather than magically minted.